zerver/tests/test_slack_importer.py

```diff
@@ -2723,3 +2729,225 @@ def test_cancel_realm_import_realm_created(self) -> None:
+                "realm": 0,
+            },
+            {
+                "id": 10,
+                "name": "unknown_content_type_and_not_guessable",
+                "file_name": None,
+                "is_animated": False,
+                "deactivated": False,
+                "author": None,
+                "realm": 0,
+            },
+        ]
+        emoji_url_map = {
+            "unknown_content_type": "https://emoji.slack-edge.com/T06NRA6HM3P/slack_zulip/3d93d672c2ba4dfc.gif",
+            "unknown_content_type_and_not_guessable": "https://emoji.slack-edge.com/T06NRA6HM3P/bowtie/f3ec6f2bb0.unknown",
+        }
+        realm_id = 0
+
+        responses.add(
+            responses.GET,
+            "https://emoji.slack-edge.com/T06NRA6HM3P/slack_zulip/3d93d672c2ba4dfc.gif",
+            body=read_test_image_file("img.png"),
+            content_type=None,
+            status=200,
+        )
+
+        responses.add(
+            responses.GET,
+            "https://emoji.slack-edge.com/T06NRA6HM3P/bowtie/f3ec6f2bb0.unknown",
+            body=read_test_image_file("img.png"),
+            content_type=None,
+            status=200,
+        )
+
+        output_dir = tempfile.mkdtemp()
+        try:
+            for emoji_data in zerver_realmemoji:
+                emoji_name = emoji_data["name"]
+                assert isinstance(emoji_name, str)
+
+                emoji_id = emoji_data["id"]
+                emoji_url = emoji_url_map[emoji_name]
+                assert isinstance(emoji_url, str)
+                assert isinstance(emoji_id, int)
+
+                ctx = (
+                    self.assertRaises(BadImageError)
+                    if emoji_name == "unknown_content_type_and_not_guessable"
+                    else nullcontext()
+                )
+                with self.assertLogs(level="WARN") as logs, ctx as e:
+                    emoji_file_name = get_emojis(
+                        output_dir,
+                        emoji_url,
+                        emoji_name,
+                        emoji_id,
+                        realm_id,
+                    )
+
+                self.assertIn(
+                    f"WARNING:root:Emoji {emoji_name} has an unspecified content type. Guessing the file extension.",
+                    logs.output,
+                )
+
+                if emoji_name == "unknown_content_type_and_not_guessable":
+                    assert e is not None
+                    self.assertIn(
+                        f"Unknown content type for: {emoji_name}",
+                        str(e.exception),
+                    )
+                    continue
+
+                assert emoji_file_name is not None
+                self.assertFalse(emoji_file_name.startswith(emoji_name))
+                self.assertRegex(emoji_file_name, r"[a-zA-Z0-9]")
+                expected_file_type = emoji_file_name.split(".")[-1]
+                self.assertTrue(emoji_file_name.endswith(expected_file_type))
```

Comment:
All of my above comments apply here, too.  I think this would be clearer if the loop were unrolled.  These particular lines are also only ever run once, anyways!